### PR TITLE
Cleanup `StringBuilder` methods

### PR DESCRIPTION
**Pull request by airblast-dev**
I haven't added the necessary tests yet so marking this as draft.

Many of `StringBuilder`'s methods had logic that really should have been their own function with safe wrappers.
This PR moves some logic to their own methods and allows public calls to allocate space without pushing bytes.

This should also reduce the effect of the memory leaks from `String` as `String::from_bytes` will now only allocate the minimal amount needed. This shouldn't change performance in any call site as the capacity is not stored, meaning no length changes are possible once a `String` is constructed. This only helps to reduce the allocated size where arguments require a `String` and a user calls `String::from_bytes`.

As part of the cleanup a redundant null check was removed from `StringBuilder::push_bytes` as `libc::realloc` already handles the null case.

I have made the commit messages to make things are easy to follow. 
Will mark this ready for review once I have added some tests.

### CODE REVIEW

**Comment by airblast-dev** [comment]
Fixed things locally, will push the latest version with new tests added and existing ones passing.

**Comment by airblast-dev** [comment]
Added a few safety checks to avoid UB in edge cases. Passes the existing tests and the new ones.

The refactor makes it so `String` either points to null with a length of 0, or points to an allocation with a length bigger than zero. The old version would allocate even when pushing 0 bytes.

Lastly, a `String` allocation is now shrunk to its minimal size when calling `StringBuilder::finish` (this is not done in `StringBuilder::finish_unchecked` for cases where extra capacity is known to not exist such as `String::from_bytes`). 

Should be ready for a review :)

File: crates/types/src/string.rs [hunk 1]
```diff
@@ -201,16 +180,154 @@ impl StringBuilder {
...
+        }
+        let remaining = self.remaining_capacity();
+        if remaining > cap {
+            return;
+        }
+        let needed_space =
+            if self.inner.data.is_null() { cap + 1 } else { cap - remaining };
+        let new_capacity = self.cap + needed_space;
+        unsafe { self.reserve_raw(new_capacity) };
+        self.cap = new_capacity;
```
Comment (noib3) [inline crates/types/src/string.rs:236]: I noticed that the bodies of `reserve` and `reserve_exact` are almost identical. Perhaps we should introduce a new helper method, something like:

```rust
/// Returns the minimum capacity needed to allocate `additional` bytes, or
/// `None` if the current capacity is already large enough.
fn min_capacity_for_additional(&self, additional: usize) -> Option<NonZeroUsize>;
```

And then:
```rust
pub fn reserve(&mut self, additional: usize) {
    let Some(min_capacity) = self.min_capacity_for_additional(additional) else {
      return;
    };
    let Some(new_capacity) = min_capacity
        .checked_next_power_of_two()
        .unwrap_or(min_capacity);
    self.reserve_raw(new_capacity);
}

pub fn reserve_exact(&mut self, additional: usize) {
    if let Some(new_capacity) = self.min_capacity_for_additional(additional) {
        self.reserve_raw(new_capacity);
    }
}
```

Which has the added benefit of making the difference between `reserve` and `reserve_exact`'s implementations much clearer at a glance.

File: crates/types/src/string.rs [hunk 2]
```diff
@@ -201,16 +180,154 @@ impl StringBuilder {
...
+    }
+
+    /// Reserve at least "by" bytes.
+    ///
+    /// This is marked unsafe as the allocation size should never be zero for portability reasons.
```
Comment (noib3) [inline crates/types/src/string.rs:241]: The function can be made safe by including this invariant in the function's signature, i.e.

```rust
fn reserve_raw(&mut self, new_capacity: NonZeroUsize)
```

File: crates/types/src/string.rs [hunk 3]
```diff
@@ -201,16 +180,154 @@ impl StringBuilder {
...
+    /// A check isn't included as all call sites should already handle the case.
+    unsafe fn reserve_raw(&mut self, by: usize) {
```
Comment (noib3) [inline crates/types/src/string.rs:243]: `by` is not very descriptive (I'm not even sure what it stands for). Maybe `new_capacity`?

File: crates/types/src/string.rs [hunk 4]
```diff
@@ -201,16 +180,154 @@ impl StringBuilder {
...
+        let ptr =
+            unsafe { libc::realloc(self.inner.data as *mut ffi::c_void, by) };
+        // realloc may return null if it is unable to allocate the requested memory
+        if ptr.is_null() {
+            unable_to_alloc_memory();
+        }
+        self.inner.data = ptr as *mut ffi::c_char;
```
Comment (noib3) [inline crates/types/src/string.rs:250]: Updating the capacity is currently left to the callers. Is there a reason for it? I think it should be this function's responsibility.

File: crates/types/src/string.rs [hunk 5]
```diff
@@ -201,16 +180,154 @@ impl StringBuilder {
...
+                };
+                if ptr.is_null() {
+                    unable_to_alloc_memory();
+                }
+                s.data = ptr as *mut ffi::c_char;
+            }
+        };
+
+        self.finish_unchecked()
+    }
```
Comment (noib3) [inline crates/types/src/string.rs:280]: Strings being leaked is a separate problem which should hopefully be resolved Soon™. I think we should just have a single `finish()` function.

File: crates/types/src/string.rs [hunk 6]
```diff
@@ -201,16 +180,154 @@ impl StringBuilder {
...
+            debug_assert!(self.cap > self.inner.len());
+        }
 
         // Prevent self's destructor from being called.
         std::mem::forget(self);
-
         s
     }
+
+    /// Returns the remaining capacity including the space for a null byte.
```
Comment (noib3) [inline crates/types/src/string.rs:308]: Shouldn't this be "excluding" instead of "including"? Although honestly it might be better to just reword it, something like:

```
/// Returns the remaining *usable* capacity, i.e. the remaining capacity minus
/// the space reserved for the null terminator.
```

**Review by noib3 — CHANGES_REQUESTED** [review]
Looks good after comments are addressed and tests pass. Thanks!

File: crates/types/src/string.rs [hunk 7]
```diff
@@ -89,8 +89,12 @@ impl String {
     #[inline]
     pub fn from_bytes(bytes: &[u8]) -> Self {
         let mut s = StringBuilder::new();
+        s.reserve_exact(bytes.len());
```
[the remark this answers is not in the document]
Comment (airblast-dev) [reply crates/types/src/string.rs:92]: Done

File: crates/types/src/string.rs (hunk 1 above)
Comment (airblast-dev) [reply crates/types/src/string.rs]: Moved the duplicate logic to its own function.

File: crates/types/src/string.rs (hunk 6 above)
Comment (airblast-dev) [reply crates/types/src/string.rs]: Yes that was a typo. I just copied over the comment you wrote here.

File: crates/types/src/string.rs (hunk 4 above)
Comment (airblast-dev) [reply crates/types/src/string.rs:237]: Nope, made it so it updates the capacity.

File: crates/types/src/string.rs [hunk 8]
```diff
@@ -88,8 +89,11 @@ impl String {
     /// by a null byte.
     #[inline]
     pub fn from_bytes(bytes: &[u8]) -> Self {
-        let mut s = StringBuilder::new();
+        let mut s = StringBuilder::with_capacity(bytes.len());
         s.push_bytes(bytes);
+
+        // no need to check if we need to shrink the allocation
+        // TODO: remove once strings are no longer leaked
```
Comment (noib3) [inline crates/types/src/string.rs:96]: I think this was about the previous method.

File: crates/types/src/string.rs [hunk 9]
```diff
@@ -201,16 +180,134 @@ impl StringBuilder {
...
+        Self {
+            inner: String { len: 0, data: ptr as *mut ffi::c_char },
+            cap: real_cap,
+        }
+    }
+
+    /// Reserve space for N more bytes.
+    ///
+    /// Does not allocate if enough space is available.
+    pub fn reserve(&mut self, aditional: usize) {
```
Comment (noib3) [inline crates/types/src/string.rs:202]: Typo.

File: crates/types/src/string.rs [hunk 10]
```diff
@@ -201,16 +180,134 @@ impl StringBuilder {
...
+        };
+        let new_capacity =
+            min_capacity.checked_next_power_of_two().unwrap_or(min_capacity);
+        self.reserve_raw(new_capacity);
+    }
+
+    /// Reserve space for exactly N more bytes.
+    ///
+    /// Does not allocate if enough space is available.
+    pub fn reserve_exact(&mut self, aditional: usize) {
```
Comment (noib3) [inline crates/types/src/string.rs:215]: Same typo.

File: crates/types/src/string.rs [hunk 11]
```diff
@@ -201,16 +180,134 @@ impl StringBuilder {
...
+        if let Some(new_capacity) = self.min_capacity_for_additional(aditional)
+        {
+            self.reserve_raw(new_capacity);
+        }
+    }
+
+    /// Allocate new_capacity bytes.
+    ///
+    /// Accepts [`NonZeroUsize`] allocation size should never be zero for portability reasons.
+    /// A check isn't included as all call sites should already handle the case.
```
Comment (noib3) [inline crates/types/src/string.rs:225]: This was only relevant to the old function's signature.

File: crates/types/src/string.rs [hunk 12]
```diff
@@ -201,16 +180,134 @@ impl StringBuilder {
...
+
+    /// Finish building the [`String`] but do not shrink the allocation
+    ///
+    /// Useful if we already know that the allocation does not have extra capacity.
+    ///
+    /// In debug mode checks if any of the following are true:
+    /// - data ptr is null with a capacity and length of zero.
+    /// - ptr points to an allocation and capacity > length.
+    /// - ptr points to an allocation with a capacity > 1 (if the capacity is 1 we are creating the
+    ///     allocation just to store a null byte, we should return null instead)
```
Comment (noib3) [inline crates/types/src/string.rs:249]: It made sense to include this info when the function was private, but now this is public and callers don't care about any of it.

File: crates/types/src/string.rs [hunk 13]
```diff
@@ -201,16 +180,134 @@ impl StringBuilder {
...
+            self.cap > self.inner.len(),
+            "allocated capacity must always be bigger than length"
+        );
+        self.cap - self.inner.len() - 1
+    }
+
+    #[inline]
+    fn min_capacity_for_additional(
+        &self,
+        aditional: usize,
```
Comment (noib3) [inline crates/types/src/string.rs:289]: Same typo as above.

File: crates/types/src/string.rs [hunk 14]
```diff
@@ -201,16 +180,134 @@ impl StringBuilder {
...
+    ) -> Option<NonZeroUsize> {
+        if aditional == 0 {
+            return None;
+        }
```
Comment (noib3) [inline crates/types/src/string.rs:293]: Is there a reason for special-casing this? It's already handled by the following check.

**Review by noib3 — CHANGES_REQUESTED** [review]

**Comment by noib3** [comment]
Note that a couple of tests are segfaulting.

**Comment by airblast-dev** [comment]
Yeah, not really sure why. Weird as nothing seems wrong in the String tests. Looking into it.

**Comment by airblast-dev** [comment]
Looked into the neovim source. Down the call chains the string is copied with `xstrdup` which doesn't handle the null ptr causing segfaults.

Not sure if it was intentional but `String::from_bytes` always allocating even for zero bytes was avoiding this problem.
This also means even `String::new` should allocate in case it is passed somewhere and it is attempted to be copied with `memcpy` or `xstrdup`...

Many of the functions work without an issue but to be safe its likely best to always allocate. For example the failing function relies on the null byte while some of them check the length, in order to even find the null byte the null pointer gets read and causes a segfault or the null pointer is used in a memcpy with mostly the same outcome. 

I originally though that it could be an issue, but `String::new()` returning a null ptr made me think this was fine.

Not really sure what to do from here.

**Comment by noib3** [comment]
Having to allocate even for empty strings kinda sucks, but it's an upstream issue. I think, at least for the time being, we should just add a comment explaining why we allocate even when not necessary.

> in order to even find the null byte the null pointer gets read

I'd have to look into it, but why does Neovim even need strings to be null-terminated when they have a `len` field?

**Comment by airblast-dev** [comment]
> I'd have to look into it, but why does Neovim even need strings to be null-terminated when they have a `len` field?

Not really sure tbh. I will look into the neovim source code again if I'm misunderstanding something or if there is an underlying bug somewhere. Regardless I am pretty sure that null pointers are a no-no. In case you want to try it out, set the minimum argument to `StringBuilder::with_capacity` to 1 inside `String::from_bytes` and all of the tests will pass.

**Comment by noib3** [comment]
I'll do some more investigation to better understand strings' usage in Neovim.

As far as this PR is concerned, I say we just remove or comment out the `if cap == 0` check in `StringBuilder::with_capacity()`, which should fix the segfaults.

**Comment by noib3** [comment]
Thanks!
